<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [3]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 


In [4]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [ ]:
# текст запроса
query_3_1 = f'''
SELECT
   COUNT(id)
FROM public.vacancies
'''

In [21]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\3851344990.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [11]:
query_3_2 = f'''
SELECT
   COUNT(id)
FROM public.employers
'''

In [22]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\3588252056.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [23]:
query_3_3 = f'''
SELECT
   COUNT(id)
FROM public.areas
'''

In [27]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\2018224393.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [16]:
query_3_4 = f'''
SELECT
   COUNT(id)
FROM public.industries
'''

In [25]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\2172479966.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    294

***

В базе HeadHanter есть 49197 вакансий от 23501 работадателей в 294 сферах деятельности, количество регионов-1362.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [37]:
query_4_1 = f'''
SELECT 
a.name,
count(v.id)
FROM
    public.VACANCIES v
JOIN public.AREAS a on v.area_id = a.id
GROUP BY 1
ORDER BY 2 DESC
'''


In [38]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\347208320.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0             Москва   5333
1    Санкт-Петербург   2851
2              Минск   2112
3        Новосибирск   2006
4             Алматы   1892
..               ...    ...
764       Тарко-Сале      1
765    Новоаннинский      1
766            Бирск      1
767           Сасово      1
768             Анна      1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [39]:
query_4_2 = f'''
SELECT 
count(v.id)
FROM
    public.VACANCIES v
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''


In [40]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\2168503709.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [45]:
query_4_3 = f'''
SELECT 
ROUND(AVG(salary_from)) salary_from,
ROUND(AVG(salary_to)) salary_to
FROM
    public.VACANCIES 
'''


In [46]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\2222372203.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



salary_from  salary_to
0      71065.0   110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [87]:
query_4_4 = f'''
SELECT 
schedule,
employment,
COUNT(id) 
FROM
    public.VACANCIES 
GROUP BY 1, 2
order by 3 DESC
'''

In [88]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\2981602351.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [89]:
query_4_5 = f'''
SELECT 
experience,
COUNT(id) 
FROM
    public.VACANCIES 
GROUP BY 1
order by 2 
'''

In [90]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\2601166264.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Всего вакансий 49197, из них в 24073 вакансиях указаны заработные платы. Средние значения для нижней и верхней границы зарплатной вилки- 71065 и 110537. Наиболее популярна пара типа рабочего графика и типа трудоустройства - полный день-полная занятость. Востребованность опыта работы в порядке возрастания: Более 6 лет — Нет опыта — От 3 до 6 лет — От 1 года до 3 лет.

In [ ]:
# выводы по детальному анализу вакансий

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [130]:
query_5_1 = f'''
SELECT 
a.name,
count(v.id)
FROM
    public.VACANCIES v
JOIN public.EMPLOYERS a on v.employer_id = a.id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 5
'''

In [131]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\1792814825.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [220]:
query_5_2 = f'''
SELECT 
a.name,
count(e.id) e_count,
count(v.id) v_count
FROM
    public.EMPLOYERS e
JOIN public.AREAS a on e.area = a.id
FULL OUTER JOIN public.VACANCIES v ON a.id = v.area_id
GROUP BY 1
HAVING count(v.id) = 0
ORDER BY 2 desc

'''


In [221]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\3341481329.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  e_count  v_count
0                                       Россия      410        0
1                                    Казахстан      207        0
2                           Московская область       75        0
3                           Краснодарский край       19        0
4                                     Беларусь       18        0
..                                         ...      ...      ...
181                                    Удобная        1        0
182                Майна (Ульяновская область)        1        0
183                            Верхнетуломский        1        0
184                                    Жаркент        1        0
185  Михайловка (Уфимский район, Башкортостан)        1        0

[186 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [231]:
query_5_3 = f'''
SELECT 
e.name,
count(distinct v.area_id)
FROM
    public.EMPLOYERS e
JOIN public.VACANCIES v on e.id=v.employer_id
GROUP BY e.id
ORDER BY 2 DESC
'''


In [232]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\2398953733.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0                                      Яндекс    181
1                                  Ростелеком    152
2                                  Спецремонт    116
3                      Поляков Денис Иванович     88
4                                   ООО ЕФИН      71
...                                       ...    ...
14901                            НПП Авиатрон      1
14902              Центр дистанционных торгов      1
14903  Городские Телекоммуникационные Системы      1
14904                       Введенский, Отель      1
14905              Группа Компаний Автоимпорт      1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [253]:
query_5_4 = f'''
SELECT 
count(e.id)
FROM
    public.EMPLOYERS e
FULL OUTER JOIN public.EMPLOYERS_INDUSTRIES i on e.id= i.employer_id
GROUP BY i.industry_id
HAVING i.industry_id IS NULL
'''

In [254]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\3114411492.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [261]:
query_5_5 = f'''
SELECT 
e.name
FROM
    public.EMPLOYERS e
JOIN public.EMPLOYERS_INDUSTRIES i on e.id= i.employer_id
GROUP BY 1
HAVING count(i.industry_id)=4
ORDER BY 1
'''

In [262]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\1650419958.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name
0        101 Интернет
1            21vek.by
2                2ГИС
3                  2К
4         4 пикселя +
...               ...
1131            ЮРИОН
1132  ЮТИП Технологии
1133           ЯКласс
1134            ЯрНео
1135    Яшин&Партнёры

[1136 rows x 1 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [284]:
query_5_6 = f'''
SELECT 
count(e.id)
FROM
    public.EMPLOYERS e
JOIN public.EMPLOYERS_INDUSTRIES ei on e.id= ei.employer_id
JOIN public.INDUSTRIES i on i.id= ei.industry_id

GROUP BY i.name
HAVING i.name='Разработка программного обеспечения'
'''

In [285]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\1169039133.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [17]:
big_city = (
    "Москва", "Санкт-Петербург", "Новосибирск", 
    "Екатеринбург", "Казань", "Нижний Новгород", 
    "Красноярск", "Челябинск", "Самара", "Уфа", 
    "Ростов-на-Дону", "Омск", "Краснодар", 
    "Воронеж", "Пермь", "Волгоград")

In [20]:
query_5_7 = f'''
(SELECT 
    a.name,
    count(*) cnt
FROM EMPLOYERS e
JOIN VACANCIES v on e.id = v.employer_id
JOIN AREAS a on a.id = v.area_id
WHERE e.name = 'Яндекс' and a.name in {big_city}
GROUP BY a.id
ORDER BY 2)
            
UNION ALL 

(SELECT 
    'TOTAL',
    COUNT (*) "Общ кол-во вакансий в городах-миллионниках"
FROM public.VACANCIES v
JOIN public.EMPLOYERS e ON e.id = v.employer_id
JOIN public.AREAS a ON a.id = v.area_id
WHERE e.name = 'Яндекс' and a.name in {big_city})            
''' 

In [21]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_5376\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0             TOTAL  485
1              Омск   21
2         Челябинск   22
3        Красноярск   23
4         Волгоград   24
5    Ростов-на-Дону   25
6            Казань   25
7             Пермь   25
8            Самара   26
9               Уфа   26
10        Краснодар   30
11          Воронеж   32
12      Новосибирск   35
13  Нижний Новгород   36
14     Екатеринбург   39
15  Санкт-Петербург   42
16           Москва   54

***

У 8419 работадаетелей не указаны сферы деятельности, а у 1135 по 4 сфер. В качестве сферы деятельности указана "Разработка программного обеспечения" у 3553 работадателей. Большое количество работадателей ищут работников за пределами своего региона в России-410 и в Казахстане-207. В базе HeadHanter работадатели с самым большим количеством вакансий: "Яндекс" и "Ростелеком". Эти же компании публикуют свои вакансии в наибольших количествах регионов. Всего "Яндекс" публикует в городах-миллиониках 485 вакансий, из них наибольшее количество в Москве-54, а наименьшее в Омске-21.

In [ ]:
# выводы по анализу работодателей

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [343]:
query_6_1 = f'''
SELECT
   count(id)
FROM public.VACANCIES
WHERE name ilike '%data%' or name ilike '%данн%' 
''' 

In [344]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\485020324.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [339]:
query_6_2 = f'''
SELECT
   count(id)
FROM public.VACANCIES
WHERE
    ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name LIKE '%ML%')
    AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL))
    AND (name ILIKE '%junior%' OR experience ILIKE '%Нет опыта%' OR employment ILIKE '%Стажировка%')
''' 


In [340]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\3641650667.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [341]:
query_6_3 = f'''
SELECT
   count(id)
FROM public.VACANCIES
WHERE
    ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name LIKE '%ML%')
    AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    )
    
    AND (key_skills ILIKE '%SQL%' or key_skills ILIKE '%postgres%')
''' 

In [342]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\1159587760.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [345]:
query_6_4 = f'''
SELECT
   count(id)
FROM public.VACANCIES
WHERE
    ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name LIKE '%ML%')
    AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    )
    
    AND (key_skills ILIKE '%python%')
''' 

In [346]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\972386973.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [371]:
query_6_5 = f'''
SELECT
   ROUND(AVG(length(key_skills) - length(REPLACE(key_skills, CHR(9), '' ))+1)::numeric, 2)
FROM public.VACANCIES
WHERE
    ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name LIKE '%ML%')
    AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    )
''' 

In [372]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\390248827.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [377]:
query_6_6 = f'''
SELECT
   ROUND(AVG(COALESCE(((salary_from + salary_to)/2), salary_from, salary_to)))
FROM public.VACANCIES
WHERE
    ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name LIKE '%ML%')
    AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    )
    AND (experience ILIKE '%от 3 до 6 лет%')
''' 

In [ ]:
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_9140\3561045109.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



round
0  243115.0

***

1771 вакансий имеют отношение к данным. Из них, для начинающих дата-сайентистов подходят 51 вакансий. В среднем в вакансиях для DS указывают 6.41 ключивых навыков. В 351 вакансиях в ключевых навыках указан Python, а в 201 - SQL или postgres. Средняя заработная плата для дата-сайентистов с опытом работы от 3 до 6 лет будет равна 243115.

In [ ]:
# выводы по предметному анализу

# Общий вывод по проекту

Так как я из республики Саха (Якутия), то проверим вакансии по РС(Я) и вакансии типа трудоустройства-удаленная работа, чтобы узнать возможность трудоустройства после окончания курса, без переезда за пределы республики.

In [22]:
query_1 = f'''
SELECT
   a.name,
   count(v.id)
FROM public.VACANCIES v
JOIN public.AREAS a on v.area_id = a.id
WHERE
    a.name ILIKE '%якут%'
GROUP BY 1

UNION ALL

SELECT
   'total',
   count(v.id)
FROM public.VACANCIES v
JOIN public.AREAS a on v.area_id = a.id
WHERE
    a.name ILIKE '%якут%'
'''
df = pd.read_sql_query(query_1, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_5376\3421513528.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_1, connection)


name  count
0                              total     66
1  Мирный (Республика Саха (Якутия))     12
2                             Якутск     54

Всего вакансий по РС(Я) 66: в городе Якутск-54, Мирный-12.

In [26]:
query_2 = f'''
SELECT
   a.name,
   count(v.id),
   v.name
FROM public.VACANCIES v
JOIN public.AREAS a on v.area_id = a.id
WHERE (v.name ilike '%data%' or v.name ilike '%данн%') 
AND a.name ILIKE '%якут%'
GROUP BY 1, 3
''' 
df = pd.read_sql_query(query_2, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_5376\4159580641.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_2, connection)


name  count                            name
0  Мирный (Республика Саха (Якутия))      1  Администратор баз данных (DBA)
1                             Якутск      1               Аналитик Big Data

Вакансии связанные с данными только 2.

In [28]:
query_3 = f'''
SELECT

   count(v.id)
FROM public.VACANCIES v
JOIN public.AREAS a on v.area_id = a.id
WHERE
    ((v.name ILIKE '%data scientist%') OR (v.name ILIKE '%data science%') OR (v.name ILIKE '%исследователь данных%')
    OR ((v.name LIKE '%ML%')
    AND (v.name NOT ILIKE '%HTML%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
    )
AND a.name ILIKE '%якут%'
'''
df = pd.read_sql_query(query_3, connection)
df


C:\Users\UserHome\AppData\Local\Temp\ipykernel_12160\3235448240.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3, connection)


count
0      0

К сожалению, вакансий по специализация Data Science по РС(Я) нет.

In [29]:
query_4 = f'''
SELECT
   v.employment,
   count(v.id)
FROM public.VACANCIES v
JOIN public.AREAS a on v.area_id = a.id
WHERE
    ((v.name ILIKE '%data scientist%') OR (v.name ILIKE '%data science%') OR (v.name ILIKE '%исследователь данных%')
    OR ((v.name LIKE '%ML%')
    AND (v.name NOT ILIKE '%HTML%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
    )
AND schedule = 'Удаленная работа'
GROUP BY 1
'''
df = pd.read_sql_query(query_4, connection)
df

C:\Users\UserHome\AppData\Local\Temp\ipykernel_12160\2404698325.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4, connection)


employment  count
0     Полная занятость    110
1     Проектная работа      2
2  Частичная занятость      3

Т.к. по РС(Я) вакансий по специальности дата-сайентиста ноль, то проверяем вакансии удаленнох работ. Предлагающие полную занятость-110, проектную работу-2, частичную занятость-3. Из этого можно сделать вывод, что трудоустроиться по специализации Data Science, не переезжая из РС(Я), возможно.

In [35]:
query_5 = f'''
SELECT
   employment,
   count(id)
FROM public.VACANCIES
WHERE
    (((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name LIKE '%ML%')
    AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL))
    AND (name ILIKE '%junior%' OR experience ILIKE '%Нет опыта%' OR employment ILIKE '%Стажировка%'))
    AND schedule = 'Удаленная работа'
GROUP BY 1  

''' 
df = pd.read_sql_query(query_5, connection)
df


C:\Users\UserHome\AppData\Local\Temp\ipykernel_12160\285358372.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5, connection)


employment  count
0     Полная занятость      5
1  Частичная занятость      1

In [5]:
connection.close()

Проверили также вакансии для начинающих в DS. И получили 6 вакансий.
Вывод: после окончания курса "Специализация Data Science" вполне возможно трудоустроиться не переезжая из РС(Я).